domain/relation/state/relation.go

```diff
@@ -472,6 +474,508 @@ func (st *State) GetRelationDetails(ctx context.Context, relationUUID corerelati
 	return result, errors.Capture(err)
 }
 
+// EnterScope indicates that the provided unit has joined the relation.
+//
+// The following error types can be expected to be returned:
+//   - [relationerrors.RelationNotFound] if the relation cannot be found.
+//   - [relationerrors.UnitNotFound] if no unit by the given name can be found
+//   - [relationerrors.RelationNotAlive] if the relation is not alive.
+//   - [relationerrors.UnitNotAlive] if the  is not alive.
+//   - [relationerrors.PotentialRelationUnitNotValid] if the unit entering scope
+//     is a subordinate and its endpoint has scope charm.ScopeContainer, but the
+//     principal application of the unit is not the application in the relation.
+func (st *State) EnterScope(
+	ctx context.Context,
+	relationUUID corerelation.UUID,
+	unitName unit.Name,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	type getUnit struct {
+		UnitUUID unit.UUID `db:"uuid"`
+		UnitName unit.Name `db:"name"`
+	}
```

Comment:
Per Ian's consistency doc, we're not using inline structures in domain states. Move to types.go